### ISSUE TITLE
Issues with timezones in Decrediton 1.3.1

### ISSUE THREAD

**Issue by rocknet** [decred/decrediton#1775]
I've seen multiple issues with timezone processing today in Decrediton 1.3.1.  The hours are not adjusted properly from UTC for me to US ET (-0500 currently but all my txns are during -0400).  When trying to troubleshoot this I've found a few issues, I'll use this txn as an example: [https://explorer.dcrdata.org/tx/380b4ed5726c00f87893c158fa0c8ec2c3cc6ef9ad8cf616a02255e797ed8826](https://explorer.dcrdata.org/tx/380b4ed5726c00f87893c158fa0c8ec2c3cc6ef9ad8cf616a02255e797ed8826)

Environment:
Win7 64Bit SP1
Today's date: November 4, 2018
US ET (we changed from UTC-4 to UTC-5 early this morning) https://www.timeanddate.com/time/zone/usa/new-york

Steps to reproduce
With Timezone in Settings set to "Local", open the Transactions -> History tab
Find transaction with the has specified above in my transaction history (though the problem is consistent with seemingly all transactions)
Note: For convenience, this is block 232673, with a timestamp of: 1524540145
https://explorer.dcrdata.org/api/block/232673/verbose?indent=true

Expected:
Considering the explorer shows this occurred at 3:22:25 AM UTC on April 24, 2018, I expect this to show 11:22:25 PM on April 23, 2018 because as of that date ET was -0400.

Actual:
I see the transaction with the hash above with a datetime of April 24, 2018 12:22:25 AM.  This is a difference of three hours, but US ET is never 3 hours from UTC.

Further step:
In Decrediton, click the transaction to see the details

Expected
The field next to "Purchased On" has the date and time

Actual
![image](https://user-images.githubusercontent.com/1266404/47972370-a5a45b80-e069-11e8-84ce-8a1a27505b3b.png)
Note: This appear to only be an issue with transaction type: "Voted"

Further step
Change the timezone to UTC
Export the transactions from the Export tab

Expected
Times are in UTC with no UTC offset specified

Actual
Times are the actual UTC time values, but have a local offset (-0400 for all transactions I have), which makes the times incorrect once the offset is added.

**Comment by rocknet** [comment]
Thanks.  I'll see if I can build master and try it out, that'd be later tonight for me.

Yes, the "Voted" transaction was from a ticket purchased inside decrediton using a stakepool.  It was the transaction that provided the staking reward, along with the ticket price back.  Each of the "Voted" transaction types I tried all had the same issue with reporting the "Purchased On" field.

**Comment by rocknet** [comment]
I think I spoke too soon. I looked at doing builds etc and I don't think I could get that going tonight. I'd need some time and likely assistance. I'm running Windows and don't have any of the decrediton dev and build tools installed yet.

I think without extra time the best I can do is test the next release. There are no nightly build binaries right?

Sent on mobile device.

On Nov 5, 2018, at 3:05 PM, Matheus Degiovani <notifications@github.com<mailto:notifications@github.com>> wrote:


Please let me know the results of the test.

It seems there's still an issue re: timezones in DST due to the current version of electron we use (2.0.8) using node 8.9.3 vs the most current node, so I'm looking whether updating to the latest electron fixes this issue (something we have to do anyway).

**Comment by matheusd** [comment]
I have weekly builds stored here: https://github.com/matheusd/decred-weekly-builds/releases

Though given that they are produced in untrusted environments and are untested, you might wanna use them on a vm or different computer.

We can give you some assistance in setting things up, but should be pretty straightforward following the readmes in the appropriate repos (dcrd/dcrwallet/decrediton). If you can do that (prepare a dev env for testing the current master branches) that would be preferable to running the weekly builds and you could help us out in testing new versions/features.

**Comment by matheusd** [comment]
I've finally been able to test using electron 3 (the current stable version) and electron 4-beta.5 (the next version still in development).

They update the internal node version and the tz database. In electron 4, the tz database version is provided in process.versions, so I can see it's using IANA's 2018.e tzdb (ftp://ftp.iana.org/tz/tzdb-2018e/). 

Electron 3 solves most issues I've had, except for a small window of time on the last couple of weeks (when using America/Sao_Paulo as my system tz) due to the rules for DST changing in Brazil this year. Electron 4 also solves that (the 2018e database already contains the rule change for this year, which electron 3, using last year's database does not have).

We'll update to electron 4 after it's out of beta, but this will be a recurring problem due to electron using a tzdb other then the system's.

**Comment by rocknet** [comment]
Glad to hear the Electron updates helped.  It is interesting that is uses an internal tzdb.  Talk about a source of bugs.

I'll try to get a dev environment running on Saturday.  I'm hoping I'll be able to use the dcrd/dcrwallet that I already have rather than build them.  I saw something about specifying the daemon process.  Also, is it a horrible idea to use mainnet for the testing?  I figure I already have problem transactions to look at in mainnet, getting testnet up would be another thing to deal with.

**Comment by matheusd** [comment]
1. You **won't** be able to use the 1.3.1 dcrd/dcrwallet with the latest decrediton master; you'll need to build dcrd/dcrwallet against their masters as well.

2. You can specify it by modifying your `config.json` file to use `"daemon_start_advanced": true`. Then you need to specify the address, credentials e rpc.cert of a running dcrd instance. This is mainly to prevent having to re-run dcrd every time you start decrediton, which is less of an issue now that the startup time for dcrd has been lowered considerably. I'd recommend leaving this off for starters.

3. Depends on the test. Just looking at the transaction history is fine, and usually there aren't show stopper bugs, but ideally you should play it safe and either use testnet (where it isn't a problem if things go wrong) or use a mainnet wallet specifically for testing (with a different seed and low amount of funds) in case you hit any snags. It's also a good idea to run these on a VM so your test wallets aren't mixed up with your production wallets.

**Comment by rocknet** [comment]
It seems like a good option for testing the timezone issues in mainnet would be for me to restore my wallet as watch only in my dev environment.  Of course getting the masterpubkey from my regular install Decrediton in Windows isn't exactly obvious to do.  I got past the 401 unauthorized by specifying the username and password I found in the dcrd.conf file, but now I get 
`open C:\Users\<username>\AppData\Local\Dcrwallet\rpc.cert: The system cannot find the file specified.`  I'm guessing Decrediton specifies something in configuration to make this work.  If I could get some help on getting the masterpubkey I can test the issues I was having.

Side note: It'd be cool if there was a console tab in Decrediton where you could execute dcrctl commands.

**Comment by matheusd** [comment]
> Lastly, in the master version of Decrediton, it still had erroneous time formats when exporting transactions while your timezone setting in Decrediton was "UTC". I fixed that locally and created a PR.

Please try with PR #1777 applied so that you can see if the timezone issues remain. Testing with electron 4 would be better, but given it's still beta, it's likely it will take a bit longer to be merged.

> Of course getting the masterpubkey from my regular install Decrediton in Windows isn't exactly obvious to do

You can export a master pub key in the new version by going to the Accounts page, opening the account then clicking on "reveal".

> Side note: It'd be cool if there was a console tab in Decrediton where you could execute dcrctl commands.

It's been discussed, but Decrediton runs with the legacy rpc disabled. The GRPC endpoints are nicer anyway :)

Btw, that particular error (the rpc.cert not being found) is because each wallet uses a separate rpc.cert file. You need to run dcrctl with `-c` to manually specify the location of the wallet's rpc.cert file.

**Comment by rocknet** [comment]
Ok, after quite a while working on getting my masterpubkey from my production wallet, I was able to import it into Decrediton built from master and all the items I mentioned at the beginning of this issue are resolved.  I tested both UTC setting and local setting, looking at transactions in UTC-0400 and -0500, exporting both to CSV, and checking the timestamp info on those older vote transactions.  All good.
